File: apps/babybox-service/internal/domain/issue.go
package domain

import (
	"time"
)

type IssueComment struct {
	Text      string    `json:"text" bson:"text"`
	Timestamp time.Time `json:"timestamp" bson:"timestamp"`
	Username  string    `bson:"username" json:"username"`
}

type IssueStateUpdate struct {
	State     string    `bson:"state" json:"state"` // Possible states: Created, Open, Denied, Solved
	Timestamp time.Time `bson:"timestamp" json:"timestamp"`
	Username  string    `bson:"username" json:"username"`
}

type BabyboxIssueDescription struct {
	Type        string `json:"type" bson:"type"`
	Subtype     string `json:"subtype" bson:"subtype"`
	Description string `json:"description,omitempty" bson:"description,omitempty"` // Optional field
	Context     string `json:"context,omitempty" bson:"context,omitempty"`         // Optional field
}

type BabyboxIssue struct {
	ID            string                  `json:"id,omitempty" bson:"_id,omitempty"`                        // Optional field
	MaintenanceID string                  `json:"maintenance_id,omitempty" bson:"maintenance_id,omitempty"` // Optional field
	Slug          string                  `json:"slug" bson:"slug"`
	Title         string                  `json:"title" bson:"title"`
	Priority      string                  `json:"priority" bson:"priority"`
	Severity      string                  `json:"severity" bson:"severity"`
	Assignee      string                  `json:"assignee,omitempty" bson:"assignee,omitempty"` // Optional field
	Issue         BabyboxIssueDescription `json:"issue" bson:"issue"`

	// Histories: first element = latest = newest; last element = earliest = oldest
	StateHistory []IssueStateUpdate `json:"state_history" bson:"state_history"`
	Comments     []IssueComment     `json:"comments,omitempty" bson:"comments,omitempty"`
}

File: apps/web/utils/issue.ts
import { BabyboxIssue } from "@/types/babybox.types";

export type BabyboxIssueHistoryPoint = {
  type: "comment" | "state";
  value: string;
  timestamp: Date;
  username: string;
};

/**
 * Goes through the histories in issue and combines them into one array representing the whole history
 * First element in the history is the last event (oldest); last element is the latest event
 */
export function combineHistories(
  issue: BabyboxIssue,
): BabyboxIssueHistoryPoint[] {
  console.log(issue);
  const states = issue.state_history.map((s) => ({
    type: "state" as "state" | "comment",
    value: s.state,
    timestamp: s.timestamp,
    username: s.username,
  }));
  const comments = issue.comments.map((s) => ({
    type: "comment" as "state" | "comment",
    value: s.text,
    timestamp: s.timestamp,
    username: s.username,
  }));
  return states.concat(comments).sort((a, b) => {
    if (a.type === "state" && a.value === "created") return -1;
    if (b.type === "state" && b.value === "created") return 1;

    return a.timestamp.getTime() - b.timestamp.getTime();
  });
}
